# Лабораторная работа 2-2

Сергеева Анастасия АДЭУ-211

## 2.2.1: Reading and Writing Data

Файл [grades.csv](grades.csv) представляет собой файл с именами учащихся и буквенными оценками:

```
Alice; A
Bob; B
Robert; A
Richard; C
```

Записать данные из файла в массив. Добавьте средний балл учащегося в строку учащегося (A=4,B=3,C=2,D=1).

Подсказка: функция [strip()](https://docs.python.org/3/library/stdtypes.html#str.strip) удаляет конечные пробелы из строки.

Запишите этот файл в новый файл `grades_gpa.csv`.

In [ ]:
import pandas as pd
#загрузка файла с компьютера
from google.colab import files
uploaded = files.upload()

Saving grades.csv to grades.csv


In [ ]:
#словарь для сопоставления буквенных оценок и их числовых значений
gpas = {"A":4, "B":3, "C":2, "D":1}

In [ ]:
# импортировать библиотеку csv
import csv

# инициализировать массив верхнего уровня
data_values = []

# откройте файл и добавьте строки в виде массивов в data_values
with open('grades.csv') as csvfile:
    filereader = csv.reader(csvfile, delimiter=';', quotechar='"')
    for row in filereader:
        # Добавление среднего балла
        row.append(gpas[row[1].strip()])
        data_values.append(row)

In [ ]:
# Создание нового DataFrame
df = pd.DataFrame(data_values, columns=['Имя', 'Буквенная оценка', 'Числовая оценка'])

# Запись данных в новый файл grades_gpa.csv
df.to_csv('grades_gpa.csv', index=False)

In [ ]:
df

Имя Буквенная оценка  Числовая оценка
0    Alice                A                4
1      Bob                B                3
2   Robert                A                4
3  Richard                C                2

## 2.2.2: Data Frames

* Рассчитайте средние сертифицированные продажи для всех альбомов.

In [3]:
import pandas as pd
#загрузка файла с компьютера
from google.colab import files
uploaded = files.upload()

Saving hit_albums.csv to hit_albums.csv


In [4]:
import pandas as pd
hit_albums = pd.read_csv("hit_albums.csv")

In [ ]:
hit_albums

Artist  \
0                     Michael Jackson   
1                               AC/DC   
2                          Pink Floyd   
3   Whitney Houston / Various artists   
4                           Meat Loaf   
..                                ...   
72                      Tracy Chapman   
73                              Usher   
74                    Various artists   
75                    Whitney Houston   
76                            Shakira   

                                                Album  Released  \
0                                            Thriller      1982   
1                                       Back in Black      1980   
2                           The Dark Side of the Moon      1973   
3                                       The Bodyguard      1992   
4                                     Bat Out of Hell      1977   
..                                                ...       ...   
72                                      Tracy Chapman      1988   
73                                        Confessions      2004   
74  Flashdance: Original Soundtrack from the Motio...      1983   
75                                            Whitney      1987   
76                                    Laundry Service      2001   

                          Genre  Certified sales (millions)  \
0                Pop, rock, R&B                        45.4   
1                     Hard rock                        25.9   
2              Progressive rock                        22.7   
3     Soundtrack/R&B, soul, pop                        27.4   
4   Hard rock, progressive rock                        20.6   
..                          ...                         ...   
72                    Folk rock                        20.0   
73                          R&B                        20.0   
74                   Soundtrack                        20.0   
75                     Pop, R&B                        20.0   
76                    Pop, Rock                        20.0   

    Claimed sales (millions)  
0                       65.0  
1                       50.0  
2                       45.0  
3                       44.0  
4                       43.0  
..                       ...  
72                       NaN  
73                       NaN  
74                       NaN  
75                       NaN  
76                       NaN  

[77 rows x 6 columns]

In [5]:
hit_albums.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77 entries, 0 to 76
Data columns (total 6 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Artist                      77 non-null     object 
 1   Album                       77 non-null     object 
 2   Released                    77 non-null     int64  
 3   Genre                       77 non-null     object 
 4   Certified sales (millions)  77 non-null     float64
 5   Claimed sales (millions)    31 non-null     float64
dtypes: float64(2), int64(1), object(3)
memory usage: 3.7+ KB


In [ ]:
#средние сертифицированные продажи для всех альбомов.
m= hit_albums['Certified sales (millions)'].mean()
m

22.40909090909091

* Создайте новый фрейм данных, содержащий только альбомы с сертифицированными продажами (более 20 миллионов человек купили альбом).

In [ ]:
#вывод альбомов с продажами больше 20 миллионов
hit_albums1 = hit_albums[hit_albums['Certified sales (millions)'] > 20]
#hit_albums1

#вывод столбцов с названием альбома и продажами
hit_albums1[['Album', 'Certified sales (millions)']]

Album  \
0                                            Thriller   
1                                       Back in Black   
2                           The Dark Side of the Moon   
3                                       The Bodyguard   
4                                     Bat Out of Hell   
5                     Their Greatest Hits (1971–1975)   
7                                             Rumours   
8                                        Come On Over   
9                                     Led Zeppelin IV   
10                                                Bad   
11                                 Jagged Little Pill   
12                                   Falling into You   
14                                   Hotel California   
19                                                  1   
20                                                 21   
29                                       Supernatural   
30                           Appetite for Destruction   
31                                Gold: Greatest Hits   
32                                  Slippery When Wet   
33                                              Spice   
34  Grease: The Original Soundtrack from the Motio...   
35                              ...Baby One More Time   
36                                      Hybrid Theory   
37       Legend: The Best of Bob Marley & The Wailers   
38                                           Tapestry   
39                                      Like a Virgin   
40                                          True Blue   
41                                           Daydream   
42                                  Come Away with Me   
43                                 No Jacket Required   
44                                      Greatest Hits   
45                         Bridge over Troubled Water   
46                                    The Joshua Tree   
47                                    Whitney Houston   
48                Backstreet's Back / Backstreet Boys   
49                                         Millennium   
50                              Happy Nation/The Sign   
51                                      CrazySexyCool   
52                                   She's So Unusual   
53                  (What's the Story) Morning Glory?   
54                                         Cross Road   
55                            The Marshall Mathers LP   

    Certified sales (millions)  
0                         45.4  
1                         25.9  
2                         22.7  
3                         27.4  
4                         20.6  
5                         32.2  
7                         27.9  
8                         29.6  
9                         29.0  
10                        20.3  
11                        24.8  
12                        20.2  
14                        21.5  
19                        21.6  
20                        22.3  
29                        20.5  
30                        21.3  
31                        29.0  
32                        28.0  
33                        28.0  
34                        28.0  
35                        28.0  
36                        27.0  
37                        25.0  
38                        25.0  
39                        25.0  
40                        25.0  
41                        25.0  
42                        25.0  
43                        25.0  
44                        25.0  
45                        25.0  
46                        25.0  
47                        25.0  
48                        24.0  
49                        24.0  
50                        23.0  
51                        23.0  
52                        22.0  
53                        22.0  
54                        21.0  
55                        21.0

Видно, что больше всего купили альбом Thriller, всего 56 альбомов смогли собрать продажи более 20 миллионов из 77.


 * Создайте новый фрейм данных на основе набора данных hit_albums, который содержит только исполнителей, у которых в списке есть как минимум два альбома.

In [60]:
#считаем количество альбомов для каждого артиста
hit_albums2=hit_albums
mask = hit_albums2["Artist"].value_counts().loc[lambda x: x >= 2].reset_index()

#переименовываем столбцы
mask.rename(columns={"index": "Artist", "Artist": "Count album"}, inplace = True)
mask

Artist  Count album
0   Michael Jackson            5
1           Madonna            3
2      Shania Twain            3
3       The Beatles            3
4   Various artists            3
5          Bon Jovi            2
6        Pink Floyd            2
7   Whitney Houston            2
8   Backstreet Boys            2
9            Eagles            2
10   Britney Spears            2
11      Celine Dion            2
12     Mariah Carey            2
13            Adele            2

In [61]:
#объединение таблиц
hit_albums3=pd.merge(hit_albums2, mask, on='Artist')

#удаление лишнего столбца
hit_albums3.drop('Count album_x', axis=1, inplace=True)

#переименование столбца
hit_albums3.rename(columns={"Count album_y": "Count album"}, inplace = True )
hit_albums3

Artist                                              Album  \
0   Michael Jackson                                           Thriller   
1   Michael Jackson                                                Bad   
2   Michael Jackson                                          Dangerous   
3   Michael Jackson          HIStory: Past, Present and Future, Book I   
4   Michael Jackson                                       Off the Wall   
5        Pink Floyd                          The Dark Side of the Moon   
6        Pink Floyd                                           The Wall   
7            Eagles                    Their Greatest Hits (1971–1975)   
8            Eagles                                   Hotel California   
9      Shania Twain                                       Come On Over   
10     Shania Twain                                    The Woman in Me   
11     Shania Twain                                                Up!   
12      Celine Dion                                   Falling into You   
13      Celine Dion                              Let's Talk About Love   
14      The Beatles              Sgt. Pepper's Lonely Hearts Club Band   
15      The Beatles                                                  1   
16      The Beatles                                         Abbey Road   
17     Mariah Carey                                          Music Box   
18     Mariah Carey                                           Daydream   
19  Various artists                                      Dirty Dancing   
20  Various artists  Grease: The Original Soundtrack from the Motio...   
21  Various artists  Flashdance: Original Soundtrack from the Motio...   
22            Adele                                                 21   
23            Adele                                                 25   
24          Madonna                          The Immaculate Collection   
25          Madonna                                      Like a Virgin   
26          Madonna                                          True Blue   
27         Bon Jovi                                  Slippery When Wet   
28         Bon Jovi                                         Cross Road   
29   Britney Spears                              ...Baby One More Time   
30   Britney Spears                            Oops!... I Did It Again   
31  Whitney Houston                                    Whitney Houston   
32  Whitney Houston                                            Whitney   
33  Backstreet Boys                Backstreet's Back / Backstreet Boys   
34  Backstreet Boys                                         Millennium   

    Released                       Genre  Certified sales (millions)  \
0       1982              Pop, rock, R&B                        45.4   
1       1987             Pop, funk, rock                        20.3   
2       1991             Rock, Funk, Pop                        17.6   
3       1995              Pop, rock, R&B                        20.0   
4       1979            Soul, disco, R&B                        20.0   
5       1973            Progressive rock                        22.7   
6       1979            Progressive rock                        17.6   
7       1976  Rock, soft rock, folk rock                        32.2   
8       1976  Rock, soft rock, folk rock                        21.5   
9       1997                Country, pop                        29.6   
10      1995                Country, pop                        20.0   
11      2002   Country, pop, world music                        20.0   
12      1996              Pop, Soft rock                        20.2   
13      1997              Pop, Soft rock                        19.3   
14      1967                        Rock                        13.1   
15      2000                        Rock                        21.6   
16      1969                        Rock                        14.4   
17      1993              Pop, R&B, Rock            

В результате вывелась таблица с артистами, у которых как минимум 2 альбома, и количеством альбомов для каждого артиста.

Больше всего альбомов у Michael Jackson (5).

* Создайте новый фрейм данных, содержащий общую сумму всех сертифицированных продаж за каждый год.

In [65]:
#группировка по году
hit_albums4 = hit_albums['Certified sales (millions)'].groupby(hit_albums['Released']).sum()

#сбрасывание индексации
hit_albums4.reset_index()

Released  Certified sales (millions)
0       1967                        13.1
1       1969                        14.4
2       1970                        25.0
3       1971                        54.0
4       1973                        22.7
5       1976                        73.7
6       1977                        67.5
7       1978                        28.0
8       1979                        57.6
9       1980                        25.9
10      1981                        25.0
11      1982                        45.4
12      1983                        62.0
13      1984                       109.6
14      1985                        67.7
15      1986                        53.0
16      1987                       144.5
17      1988                        20.0
18      1990                        19.4
19      1991                        54.2
20      1992                        76.4
21      1993                        42.0
22      1994                        64.0
23      1995                       111.8
24      1996                        48.2
25      1997                        91.0
26      1999                        72.5
27      2000                        89.6
28      2001                        20.0
29      2002                        65.0
30      2004                        20.0
31      2011                        22.3
32      2015                        20.0

Больше всего продаж (144.5) в 1987 году, что обусловлено выходом альбомов сразу нескольких известных личностей, одна из которых самая продаваемая по проведенному выше анализу (Michael Jackson	), к тому же в тот период Европа переживала расцвет дэнс-поп музыки. Меньше всего продаж в 1967 году (13.1), ведь вышел только 1 альбом.